dex/networks/eth/contracts/ETHSwapV1.sol

```diff
@@ -0,0 +1,231 @@
+// SPDX-License-Identifier: BlueOak-1.0.0
+// pragma should be as specific as possible to allow easier validation.
+pragma solidity = 0.8.18;
+
+// ETHSwap creates a contract to be deployed on an ethereum network. After
+// deployed, it keeps a record of the state of a contract and enables
+// redemption and refund of the contract when conditions are met.
+//
+// ETHSwap accomplishes this by holding funds sent to ETHSwap until certain
+// conditions are met. An initiator sends a tx with the Vector(s) to fund and
+// the requisite value to transfer to ETHSwap. At
+// this point the funds belong to the contract, and cannot be accessed by
+// anyone else, not even the contract's deployer. The swap Vector specifies
+// the conditions necessary for refund and redeem.
+//
+// ETHSwap has no limits on gas used for any transactions.
+//
+// ETHSwap cannot be used by other contracts or by a third party mediating
+// the swap or multisig wallets.
+//
+// This code should be verifiable as resulting in a certain on-chain contract
+// by compiling with the correct version of solidity and comparing the
+// resulting byte code to the data in the original transaction.
+contract ETHSwap {
+    // Step is a type that hold's a contract's current step. Empty is the
+    // uninitiated or null value.
+    enum Step { Empty, Filled, Redeemed, Refunded }
+
+    struct Status {
+        Step step;
+        bytes32 secret;
+        uint256 blockNumber;
+    }
+
+    bytes32 constant RefundRecord = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
+
+    // swaps is a map of contract hashes to the "swap record". The swap record
+    // has the following interpretation.
+    //   if (record == bytes32(0x00)): contract is uninitiated
+    //   else if (uint256(record) < block.number && sha256(record) != contract.secretHash):
+    //      contract is initiated and redeemable by the participant with the secret.
+    //   else if (sha256(record) == contract.secretHash): contract has been redeemed
+    //   else if (record == RefundRecord): contract has been refunded
+    //   else: invalid record. Should be impossible by construction
+    mapping(bytes32 => bytes32) public swaps;
+
+    // Vector is the information necessary for initialization and redemption
+    // or refund. The Vector itself is not stored on-chain. Instead, a key
+    // unique to the Vector is generated from the Vector data and keys
+    // the swap record.
+    struct Vector {
+        bytes32 secretHash;
+        uint256 value;
+        address initiator;
+        uint64 refundTimestamp;
+        address participant;
+    }
+
+    // contractKey generates a key hash which commits to the contract data. The
+    // generated hash is used as a key in the swaps map.
+    function contractKey(Vector calldata v) public pure returns (bytes32) {
+        return sha256(bytes.concat(v.secretHash, bytes20(v.initiator), bytes20(v.participant), bytes32(v.value), bytes8(v.refundTimestamp)));
```

Comment:
I was originally using `abi.encodePacked`, but we decided to use `bytes.encode` because of some ambiguity in the implementation of abi encoding, iirc. I can't find the convo. I think it was on Matrix.

Reply:
I still need to check out keccak